PyObject value in Result can cause JepException

```
19/03/06 17:55:06 ERROR SocketSession: IllegalStateException: jep.JepException: Invalid thread access.
polynote.kernel.RuntimeError: IllegalStateException: jep.JepException: Invalid thread access.
	at polynote.kernel.ErrorResult$.apply(Result.scala:136)
	at polynote.server.SocketSession$$anonfun$toResponse$1$$anonfun$4.apply(SocketSession.scala:89)
	at polynote.server.SocketSession$$anonfun$toResponse$1$$anonfun$4.apply(SocketSession.scala:88)
	at fs2.Stream$$anonfun$handleErrorWith$extension$1.apply(Stream.scala:1324)
	at fs2.Stream$$anonfun$handleErrorWith$extension$1.apply(Stream.scala:1324)
	at fs2.internal.FreeC$$anonfun$handleErrorWith$1.apply(FreeC.scala:54)
	at fs2.internal.FreeC$$anonfun$handleErrorWith$1.apply(FreeC.scala:51)
	at fs2.internal.FreeC$ViewL$.mk(FreeC.scala:214)
	at fs2.internal.FreeC$ViewL$.apply(FreeC.scala:206)
	at fs2.internal.FreeC.viewL(FreeC.scala:67)
	at fs2.internal.Algebra$.fs2$internal$Algebra$$go$1(Algebra.scala:220)
	at fs2.internal.Algebra$$anonfun$fs2$internal$Algebra$$closeAndGo$1$1$$anonfun$apply$10.apply(Algebra.scala:319)
	at fs2.internal.Algebra$$anonfun$fs2$internal$Algebra$$closeAndGo$1$1$$anonfun$apply$10.apply(Algebra.scala:317)
	at cats.effect.internals.IORunLoop$.cats$effect$internals$IORunLoop$$loop(IORunLoop.scala:139)
	at cats.effect.internals.IORunLoop$RestartCallback.signal(IORunLoop.scala:351)
	at cats.effect.internals.IORunLoop$RestartCallback.apply(IORunLoop.scala:372)
	at cats.effect.internals.IORunLoop$RestartCallback.apply(IORunLoop.scala:312)
	at cats.effect.internals.IORunLoop$.cats$effect$internals$IORunLoop$$loop(IORunLoop.scala:136)
	at cats.effect.internals.IORunLoop$RestartCallback.signal(IORunLoop.scala:351)
	at cats.effect.internals.IORunLoop$RestartCallback.apply(IORunLoop.scala:372)
	at cats.effect.internals.IORunLoop$RestartCallback.apply(IORunLoop.scala:312)
	at cats.effect.internals.Callback$AsyncIdempotentCallback.run(Callback.scala:127)
	at cats.effect.internals.Trampoline.cats$effect$internals$Trampoline$$immediateLoop(Trampoline.scala:70)
	at cats.effect.internals.Trampoline.startLoop(Trampoline.scala:36)
	at cats.effect.internals.TrampolineEC$JVMTrampoline.cats$effect$internals$TrampolineEC$JVMTrampoline$$super$startLoop(TrampolineEC.scala:93)
	at cats.effect.internals.TrampolineEC$JVMTrampoline$$anonfun$startLoop$1.apply$mcV$sp(TrampolineEC.scala:93)
	at cats.effect.internals.TrampolineEC$JVMTrampoline$$anonfun$startLoop$1.apply(TrampolineEC.scala:93)
	at cats.effect.internals.TrampolineEC$JVMTrampoline$$anonfun$startLoop$1.apply(TrampolineEC.scala:93)
	at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:72)
	at cats.effect.internals.TrampolineEC$JVMTrampoline.startLoop(TrampolineEC.scala:92)
	at cats.effect.internals.Trampoline.execute(Trampoline.scala:43)
	at cats.effect.internals.TrampolineEC.execute(TrampolineEC.scala:44)
	at cats.effect.internals.Callback$AsyncIdempotentCallback.apply(Callback.scala:133)
	at cats.effect.internals.Callback$AsyncIdempotentCallback.apply(Callback.scala:120)
	at cats.effect.concurrent.Deferred$ConcurrentDeferred$$anonfun$2.apply(Deferred.scala:205)
	at cats.effect.concurrent.Deferred$ConcurrentDeferred$$anonfun$2.apply(Deferred.scala:205)
	at cats.effect.concurrent.Deferred$ConcurrentDeferred$$anonfun$1.apply$mcV$sp(Deferred.scala:241)
	at cats.effect.concurrent.Deferred$ConcurrentDeferred$$anonfun$1.apply(Deferred.scala:241)
	at cats.effect.concurrent.Deferred$ConcurrentDeferred$$anonfun$1.apply(Deferred.scala:241)
	at cats.effect.internals.IORunLoop$.cats$effect$internals$IORunLoop$$loop(IORunLoop.scala:87)
	at cats.effect.internals.IORunLoop$RestartCallback.signal(IORunLoop.scala:351)
	at cats.effect.internals.IORunLoop$RestartCallback.apply(IORunLoop.scala:372)
	at cats.effect.internals.IORunLoop$RestartCallback.apply(IORunLoop.scala:312)
	at cats.effect.internals.IOShift$Tick.run(IOShift.scala:36)
	at scala.concurrent.impl.ExecutionContextImpl$AdaptedForkJoinTask.exec(ExecutionContextImpl.scala:121)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
Caused by: java.lang.IllegalStateException: jep.JepException: Invalid thread access.
	at jep.python.PyObject.checkValidRuntime(PyObject.java:112)
	at jep.python.PyObject.toString(PyObject.java:679)
	at polynote.kernel.util.KernelContext$$anonfun$5.apply(KernelContext.scala:85)
	at polynote.kernel.util.KernelContext$$anonfun$5.apply(KernelContext.scala:85)
	at scala.Option.map(Option.scala:146)
	at polynote.kernel.util.KernelContext.reprsOf(KernelContext.scala:85)
	at polynote.kernel.ResultValue$.apply(Result.scala:175)
	at polynote.kernel.util.RuntimeSymbolTable$RuntimeValue.toResultValue(RuntimeSymbolTable.scala:112)
	at polynote.kernel.PolyKernel$$anonfun$currentSymbols$1$$anonfun$apply$38.apply(PolyKernel.scala:204)
	at polynote.kernel.PolyKernel$$anonfun$currentSymbols$1$$anonfun$apply$38.apply(PolyKernel.scala:204)
	at scala.collection.immutable.List.map(List.scala:288)
	at polynote.kernel.PolyKernel$$anonfun$currentSymbols$1.apply(PolyKernel.scala:204)
	at polynote.kernel.PolyKernel$$anonfun$currentSymbols$1.apply(PolyKernel.scala:204)
	at cats.effect.internals.IORunLoop$.cats$effect$internals$IORunLoop$$loop(IORunLoop.scala:87)
	at cats.effect.internals.IORunLoop$RestartCallback.signal(IORunLoop.scala:351)
	at cats.effect.internals.IORunLoop$RestartCallback.apply(IORunLoop.scala:372)
	at cats.effect.internals.IORunLoop$RestartCallback.apply(IORunLoop.scala:312)
	at cats.effect.internals.IORunLoop$.cats$effect$internals$IORunLoop$$loop(IORunLoop.scala:136)
	at cats.effect.internals.IORunLoop$.startCancelable(IORunLoop.scala:41)
	at cats.effect.internals.IOBracket$BracketStart.run(IOBracket.scala:86)
	at cats.effect.internals.Trampoline.cats$effect$internals$Trampoline$$immediateLoop(Trampoline.scala:70)
	at cats.effect.internals.Trampoline.startLoop(Trampoline.scala:36)
	at cats.effect.internals.TrampolineEC$JVMTrampoline.cats$effect$internals$TrampolineEC$JVMTrampoline$$super$startLoop(TrampolineEC.scala:93)
	at cats.effect.internals.TrampolineEC$JVMTrampoline$$anonfun$startLoop$1.apply$mcV$sp(TrampolineEC.scala:93)
	at cats.effect.internals.TrampolineEC$JVMTrampoline$$anonfun$startLoop$1.apply(TrampolineEC.scala:93)
	at cats.effect.internals.TrampolineEC$JVMTrampoline$$anonfun$startLoop$1.apply(TrampolineEC.scala:93)
	at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:72)
	at cats.effect.internals.TrampolineEC$JVMTrampoline.startLoop(TrampolineEC.scala:92)
	at cats.effect.internals.Trampoline.execute(Trampoline.scala:43)
	at cats.effect.internals.TrampolineEC.execute(TrampolineEC.scala:44)
	at cats.effect.internals.IOBracket$BracketStart.apply(IOBracket.scala:72)
	at cats.effect.internals.IOBracket$BracketStart.apply(IOBracket.scala:52)
	at cats.effect.internals.IORunLoop$.cats$effect$internals$IORunLoop$$loop(IORunLoop.scala:136)
	at cats.effect.internals.IORunLoop$.start(IORunLoop.scala:34)
	at cats.effect.internals.IOBracket$$anonfun$apply$1.apply(IOBracket.scala:44)
	at cats.effect.internals.IOBracket$$anonfun$apply$1.apply(IOBracket.scala:34)
	at cats.effect.internals.IORunLoop$RestartCallback.start(IORunLoop.scala:337)
	at cats.effect.internals.IORunLoop$.cats$effect$internals$IORunLoop$$loop(IORunLoop.scala:119)
	at cats.effect.internals.IORunLoop$.start(IORunLoop.scala:34)
	at cats.effect.internals.IOBracket$$anonfun$apply$1.apply(IOBracket.scala:44)
	at cats.effect.internals.IOBracket$$anonfun$apply$1.apply(IOBracket.scala:34)
	at cats.effect.internals.IORunLoop$RestartCallback.start(IORunLoop.scala:337)
	at cats.effect.internals.IORunLoop$.cats$effect$internals$IORunLoop$$loop(IORunLoop.scala:119)
	at cats.effect.internals.IORunLoop$RestartCallback.signal(IORunLoop.scala:351)
	at cats.effect.internals.IORunLoop$RestartCallback.apply(IORunLoop.scala:372)
	at cats.effect.internals.IORunLoop$RestartCallback.apply(IORunLoop.scala:312)
	at cats.effect.internals.IOShift$Tick.run(IOShift.scala:36)
	at scala.concurrent.impl.ExecutionContextImpl$AdaptedForkJoinTask.exec(ExecutionContextImpl.scala:121)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.pollAndExecAll(ForkJoinPool.java:1253)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1346)
	... 2 more
Caused by: jep.JepException: Invalid thread access.
	at jep.Jep.isValidThread(Jep.java:319)
	at jep.python.PyObject.checkValid(PyObject.java:94)
	at jep.python.PyObject.checkValidRuntime(PyObject.java:110)
	... 52 more
```

--- a/polynote-kernel/src/main/scala/polynote/kernel/Result.scala
+++ b/polynote-kernel/src/main/scala/polynote/kernel/Result.scala
@@ -178,6 +178,11 @@ object ResultValue extends ResultCompanion[ResultValue](4) {
   def apply(ctx: KernelContext, name: String, value: Any, sourceCell: CellID): ResultValue =
     apply(ctx, name, ctx.inferType(value), value, sourceCell)
 
+  def apply(ctx: KernelContext, name: String, reprs: TinyList[ValueRepr], typ: Universe#Type, value: Any, sourceCell: CellID): ResultValue = {
+    val (_, typeStr) = toGlobalType(ctx, typ)
+    ResultValue(name, typeStr, reprs, sourceCell, value, typ)
+  }
+
 
   // manual codec - we'll never encode nor decode `value` nor `scalaType`.
   /*

--- a/polynote-kernel/src/main/scala/polynote/kernel/util/RuntimeSymbolTable.scala
+++ b/polynote-kernel/src/main/scala/polynote/kernel/util/RuntimeSymbolTable.scala
@@ -28,7 +28,7 @@ final class RuntimeSymbolTable(
 
   private val cellIds: mutable.TreeSet[CellID] = new mutable.TreeSet()
 
-  private val newSymbols: Topic[IO, RuntimeValue] = {
+  private val newSymbols: Topic[IO, RuntimeValue] =
     Topic[IO, RuntimeValue]{
       // note: IntelliJ doesn't like this but it compiles fine...
       val kernel = RuntimeValue(
@@ -43,7 +43,7 @@ final class RuntimeSymbolTable(
       putValue(kernel)
       kernel
     }.unsafeRunSync()
-  }
+
 
   private val awaitingDelivery = SignallingRef[IO, Int](0).unsafeRunSync()
 

--- a/polynote-kernel/src/main/scala/polynote/kernel/util/RuntimeSymbolTable.scala
+++ b/polynote-kernel/src/main/scala/polynote/kernel/util/RuntimeSymbolTable.scala
@@ -8,6 +8,7 @@ import fs2.concurrent.{SignallingRef, Topic}
 import polynote.kernel.{KernelStatusUpdate, ResultValue}
 import polynote.kernel.lang.LanguageInterpreter
 import polynote.messages.CellID
+import polynote.runtime.ValueRepr
 
 import scala.collection.mutable
 import scala.collection.JavaConverters._
@@ -29,7 +30,7 @@ final class RuntimeSymbolTable(
 
   private val newSymbols: Topic[IO, RuntimeValue] =
     Topic[IO, RuntimeValue]{
-      val kernel = RuntimeValue("kernel", polynote.runtime.Runtime, global.typeOf[polynote.runtime.Runtime.type], None, -1)
+      val kernel = RuntimeValue("kernel", polynote.runtime.Runtime, None, -1)
       // make sure this is actually in the symbol table.
       // TODO: is there a better way to set this value?
       putValue(kernel)
@@ -64,7 +65,8 @@ final class RuntimeSymbolTable(
   }
 
   def publish(source: LanguageInterpreter[IO], sourceCellId: CellID)(name: String, value: Any, staticType: Option[global.Type]): IO[Unit] = {
-    val rv = RuntimeValue(name, value, staticType.getOrElse(kernelContext.inferType(value)), Some(source), sourceCellId)
+    val typ = staticType.getOrElse(kernelContext.inferType(value))
+    val rv = RuntimeValue(name, value, kernelContext.reprsOf(value, typ), typ, Some(source), sourceCellId)
     for {
       _    <- IO(putValue(rv))
       subs <- newSymbols.subscribers.head.compile.lastOrError
@@ -95,6 +97,7 @@ final class RuntimeSymbolTable(
   sealed case class RuntimeValue(
     name: String,
     value: Any,
+    reprs: List[ValueRepr],
     scalaTypeHolder: global.Type,
     source: Option[LanguageInterpreter[IO]],
     sourceCellId: CellID
@@ -109,12 +112,12 @@ final class RuntimeSymbolTable(
     // don't need to hash everything to determine hash code; name collisions are less common than hash comparisons
     override def hashCode(): Int = name.hashCode()
 
-    def toResultValue: ResultValue = ResultValue(kernelContext, name, scalaTypeHolder, value, sourceCellId)
+    def toResultValue: ResultValue = ResultValue(kernelContext, name, reprs, scalaTypeHolder, value, sourceCellId)
   }
 
   object RuntimeValue {
     def apply(name: String, value: Any, source: Option[LanguageInterpreter[IO]], sourceCell: CellID): RuntimeValue = RuntimeValue(
-      name, value, kernelContext.inferType(value), source, sourceCell
+      name, value, kernelContext.reprsOf(value, kernelContext.inferType(value)), kernelContext.inferType(value), source, sourceCell
     )
 
     def fromSymbolDecl(symbolDecl: SymbolDecl[IO]): Option[RuntimeValue] = {
@@ -126,7 +129,7 @@ final class RuntimeSymbolTable(
     def fromResultValue(resultValue: ResultValue, source: LanguageInterpreter[IO]): Option[RuntimeValue] = resultValue match {
       case ResultValue(_, _, _, _, Unit, _) => None
       case ResultValue(name, typeName, reprs, sourceCell, value, scalaType) =>
-        Some(apply(name, value, scalaType.asInstanceOf[global.Type], Option(source), sourceCell))
+        Some(apply(name, value, reprs, scalaType.asInstanceOf[global.Type], Option(source), sourceCell))
     }
   }
 

--- a/polynote-kernel/src/main/scala/polynote/kernel/util/RuntimeSymbolTable.scala
+++ b/polynote-kernel/src/main/scala/polynote/kernel/util/RuntimeSymbolTable.scala
@@ -8,7 +8,7 @@ import fs2.concurrent.{SignallingRef, Topic}
 import polynote.kernel.{KernelStatusUpdate, ResultValue}
 import polynote.kernel.lang.LanguageInterpreter
 import polynote.messages.CellID
-import polynote.runtime.ValueRepr
+import polynote.runtime.{StringRepr, ValueRepr}
 
 import scala.collection.mutable
 import scala.collection.JavaConverters._
@@ -28,14 +28,22 @@ final class RuntimeSymbolTable(
 
   private val cellIds: mutable.TreeSet[CellID] = new mutable.TreeSet()
 
-  private val newSymbols: Topic[IO, RuntimeValue] =
+  private val newSymbols: Topic[IO, RuntimeValue] = {
     Topic[IO, RuntimeValue]{
-      val kernel = RuntimeValue("kernel", polynote.runtime.Runtime, None, -1)
+      // note: IntelliJ doesn't like this but it compiles fine...
+      val kernel = RuntimeValue(
+        "kernel",
+        polynote.runtime.Runtime,
+        List(StringRepr(polynote.runtime.Runtime.getClass.getSimpleName)),
+        global.typeOf[polynote.runtime.Runtime.type],
+        None,
+        -1)
       // make sure this is actually in the symbol table.
       // TODO: is there a better way to set this value?
       putValue(kernel)
       kernel
     }.unsafeRunSync()
+  }
 
   private val awaitingDelivery = SignallingRef[IO, Int](0).unsafeRunSync()
 
@@ -116,9 +124,10 @@ final class RuntimeSymbolTable(
   }
 
   object RuntimeValue {
-    def apply(name: String, value: Any, source: Option[LanguageInterpreter[IO]], sourceCell: CellID): RuntimeValue = RuntimeValue(
-      name, value, kernelContext.reprsOf(value, kernelContext.inferType(value)), kernelContext.inferType(value), source, sourceCell
-    )
+    def apply(name: String, value: Any, source: Option[LanguageInterpreter[IO]], sourceCell: CellID): RuntimeValue = {
+      val typ = kernelContext.inferType(value)
+      RuntimeValue(name, value, kernelContext.reprsOf(value, typ), typ, source, sourceCell)
+    }
 
     def fromSymbolDecl(symbolDecl: SymbolDecl[IO]): Option[RuntimeValue] = {
       symbolDecl.getValue.map { value =>

